test/model/typing_status_test.dart

```diff
@@ -0,0 +1,240 @@
+    });
+
+    test('add a typist in the same narrow', () {
+      prepareModel();
+
+      checkTyping(groupNarrow, TypingOp.start, eg.selfUser)
+        .typistsByNarrow.deepEquals({
+          groupNarrow: {eg.selfUser.userId: anyTimer}});
+      checkNotifiedOnce();
+
+      checkTyping(groupNarrow, TypingOp.start, eg.thirdUser)
+        .typistsByNarrow.deepEquals({
+          groupNarrow: {eg.selfUser.userId: anyTimer, eg.thirdUser.userId: anyTimer}});
+      checkNotifiedOnce();
+    });
+
+    test('sort dm recipients', () {
+      prepareModel();
+
+      final eventUnsorted = TypingEvent(id: 1, op: TypingOp.start, senderId: eg.selfUser.userId,
+        messageType: MessageType.private,
+        recipientIds: [5, 4, 10, 8, 2, 1, eg.selfUser.userId],
+        streamId: null,
+        topic: null);
+      // DmNarrow's constructor expects the recipient IDs to be sorted.
+      check(() => model.handleTypingEvent(eventUnsorted)).returnsNormally();
+    });
+
+  });
+
+  group('handle typing stop events', () {
+    test('remove a typist from an unknown narrow', () => awaitFakeAsync((async) async {
+      prepareModel();
+
+      checkTyping(groupNarrow, TypingOp.start, eg.otherUser)
+        .typistsByNarrow.deepEquals({groupNarrow: {eg.otherUser.userId: anyTimer}});
+      checkNotifiedOnce();
+
+      checkTyping(dmNarrow, TypingOp.stop, eg.otherUser)
+        .typistsByNarrow.deepEquals({groupNarrow: {eg.otherUser.userId: anyTimer}});
+      checkNotNotified();
+    }));
+
+    test('remove one of two typists in the same narrow', () => awaitFakeAsync((async) async {
+      prepareModel();
+
+      checkTyping(groupNarrow, TypingOp.start, eg.otherUser);
+      checkNotifiedOnce();
+      checkTyping(groupNarrow, TypingOp.start, eg.selfUser);
+      checkNotifiedOnce();
+
+      check(model).typistsByNarrow.deepEquals({
+        groupNarrow: {eg.otherUser.userId: anyTimer, eg.selfUser.userId: anyTimer}});
+
+      checkTyping(groupNarrow, TypingOp.stop, eg.otherUser)
+        .typistsByNarrow.deepEquals({groupNarrow: {eg.selfUser.userId: anyTimer}});
+      checkNotifiedOnce();
+    }));
+
+    test('remove typists from different narrows', () => awaitFakeAsync((async) async {
+      prepareModel();
+
+      checkTyping(dmNarrow, TypingOp.start, eg.otherUser);
+      checkNotifiedOnce();
+      checkTyping(groupNarrow, TypingOp.start, eg.thirdUser);
+      checkNotifiedOnce();
+      checkTyping(topicNarrow, TypingOp.start, eg.selfUser);
+      checkNotifiedOnce();
+
+      async.elapse(const Duration(seconds: 5));
+      check(model).typistsByNarrow.deepEquals({
+        dmNarrow: {eg.otherUser.userId: anyTimer},
+        groupNarrow: {eg.thirdUser.userId: anyTimer},
+        topicNarrow: {eg.selfUser.userId: anyTimer}});
+
+      checkTyping(groupNarrow, TypingOp.stop, eg.thirdUser)
+        .typistsByNarrow.deepEquals({
+          dmNarrow: {eg.otherUser.userId: anyTimer},
+          topicNarrow: {eg.selfUser.userId: anyTimer}});
```

Comment:
One thing about this style of writing the checks is that it feels like it's making these tests a bit too closely tied to the exact data structure that `TypingStatus` uses internally. In particular, to the fact that there are all these timers there.

Relatedly, the fact that these are timers isn't actually relevant to these tests — they never do anything with the timers. (There are other tests lower in the file that test the timer behavior.) So that detail is only a distraction in these tests.

I think one thing that would clean this up further is to let the main Map data structure be private, and expose something like `Iterable<SendableNarrow> get debugActiveNarrows => typistIdsByNarrow.keys`. Then the test code can combine that debug getter with `getTypistIdsInNarrow` to get a map from narrow to iterable of sender IDs. And these checks in the individual test cases can look like:
```suggestion
      checkTypists({
          dmNarrow: [eg.otherUser],
          topicNarrow: [eg.selfUser]});
```

Because there'd be a test helper involved, it can also take care of mapping user objects to user IDs, further simplifying the individual test cases.